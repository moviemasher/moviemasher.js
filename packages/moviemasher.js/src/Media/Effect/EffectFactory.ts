import { DefinitionType } from "../../Setup/Enums"
import { EffectDefinitionClass } from "./EffectDefinitionClass"
import { Effect, EffectDefinition, EffectObject, EffectDefinitionObject } from "./Effect"
import { assertPopulatedString } from "../../Utility/Is"

import effectBlurJson from "../../MediaObjects/effect/blur.json"
import effectChromaKeyJson from "../../MediaObjects/effect/chromakey.json"
import effectEmbossJson from "../../MediaObjects/effect/emboss.json"
import effectGrayscaleJson from "../../MediaObjects/effect/grayscale.json"
import effectSepiaJson from "../../MediaObjects/effect/sepia.json"
import effectSharpenJson from "../../MediaObjects/effect/sharpen.json"
import { MediaFactories } from "../MediaFactories"
import { MediaDefaults } from "../MediaDefaults"

export const effectDefinition = (object : EffectDefinitionObject) : EffectDefinition => {
  const { id } = object
  assertPopulatedString(id)
  return new EffectDefinitionClass({...object, type: DefinitionType.Effect })
}

export const effectDefinitionFromId = (id: string): EffectDefinition => {
  const definition = MediaDefaults[DefinitionType.Effect].find(definition => 
    definition.id === id
  )
  if (definition) return definition as EffectDefinition

  return effectDefinition({ id })
}

export const effectInstance = (object: EffectObject): Effect => {
  const { definitionId = '' } = object
  // console.log("effectInstance", definitionId, object)
  const definition = effectDefinitionFromId(definitionId)
  return definition.instanceFromObject(object)
}

export const effectFromId = (definitionId: string): Effect => {
  const definition = effectDefinitionFromId(definitionId)
  return definition.instanceFromObject()
}
MediaFactories[DefinitionType.Effect] = effectDefinition
MediaDefaults[DefinitionType.Effect].push(
  effectDefinition(effectBlurJson),
  effectDefinition(effectChromaKeyJson),
  effectDefinition(effectEmbossJson),
  effectDefinition(effectGrayscaleJson),
  effectDefinition(effectSepiaJson),
  effectDefinition(effectSharpenJson),
)