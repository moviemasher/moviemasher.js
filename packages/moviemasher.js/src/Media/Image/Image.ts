import { Container, ContainerDefinition } from "../../Container/Container"
import { Content, ContentDefinition, ContentDefinitionObject, ContentObject } from "../../Content/Content"
import { GenericFactory } from "../../declarations"
import { FilterDefinition } from "../../Filter/Filter"
import {
  UpdatableDimensions, UpdatableDimensionsDefinition, UpdatableDimensionsDefinitionObject, UpdatableDimensionsObject
} from "../../Mixin/UpdatableDimensions/UpdatableDimensions"

export interface ImageObject extends ContentObject, UpdatableDimensionsObject {}

export interface ImageDefinitionObject extends ContentDefinitionObject, UpdatableDimensionsDefinitionObject {}

export interface Image extends Content, Container, UpdatableDimensions {
  definition : ImageDefinition
}

export interface ImageDefinition extends ContainerDefinition, ContentDefinition, UpdatableDimensionsDefinition {
  instanceFromObject(object?: ImageObject): Image
  setptsFilterDefinition: FilterDefinition
}

/**
 * @category Factory
 */
export interface ImageFactory extends GenericFactory<Image, ImageObject, ImageDefinition, ImageDefinitionObject> {}
