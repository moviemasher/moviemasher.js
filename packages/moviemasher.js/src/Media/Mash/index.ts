export * from "./Mash"
export * from "./MashMedia"