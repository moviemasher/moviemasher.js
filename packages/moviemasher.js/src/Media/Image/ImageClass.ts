import { LoadedImage, SvgContent } from "../../declarations"
import { Dimensions } from "../../Setup/Dimensions"
import { Chain, ContainerChainArgs, ContentChainArgs, GraphFile } from "../../MoveMe"
import { LoadType, Phase } from "../../Setup/Enums"
import { InstanceBase } from "../../Instance/InstanceBase"
import { ImageDefinition, Image } from "./Image"
import { FilterChain } from "../../Edited/Mash/FilterChain/FilterChain"
import { assertPopulatedString, isAboveZero } from "../../Utility/Is"
import { PreloadableMixin } from "../../Mixin/Preloadable/PreloadableMixin"
import { UpdatableDimensionsMixin } from "../../Mixin/UpdatableDimensions/UpdatableDimensionsMixin"
import { ChainLinks, Filter, FilterChainPhase } from "../../Filter/Filter"
import { TrackPreview } from "../../Editor/Preview/TrackPreview/TrackPreview"
import { ContentMixin } from "../../Content/ContentMixin"
import { ContainerMixin } from "../../Container/ContainerMixin"


const ImageWithContainer = ContainerMixin(InstanceBase)
const ImageWithContent = ContentMixin(ImageWithContainer)
const ImageWithPreloadable = PreloadableMixin(ImageWithContent)
const ImageWithUpdatable = UpdatableDimensionsMixin(ImageWithPreloadable)
export class ImageClass extends ImageWithUpdatable implements Image {
  constructor(...args: any[]) {
    super(...args)
    this.setptsFilter = this.definition.setptsFilterDefinition.instanceFromObject()
  }

  contentChain(args: ContentChainArgs): Chain {
    const chain = super.contentChain(args)
    const { commandFiles} = chain
    const graphFile = this.graphFile(args.editing)
    const commandFile = { ...graphFile, inputId: this.id }
    commandFiles.push(commandFile)
    return chain
  }

  containerChain(args: ContainerChainArgs): Chain {
    const chain = super.containerChain(args)
    const { commandFiles} = chain
    const graphFile = this.graphFile(args.editing)
    const commandFile = { ...graphFile, inputId: this.id }
    commandFiles.push(commandFile)
    return chain
  }

  chainLinks(): ChainLinks {
    const links: ChainLinks = []
    links.push(this, this.setptsFilter)
    links.push(...super.chainLinks())
    return links
  }

  declare definition: ImageDefinition

  filterChainPhase(filterChain: FilterChain, phase: Phase): FilterChainPhase | undefined {
    if (phase !== Phase.Initialize) return

    const { filterGraph } = filterChain
    const { streaming, editing, preloader } = filterGraph
    const graphFile = this.graphFile(editing)
    let { width, height } = this.definition
    if (!(isAboveZero(width) && isAboveZero(height))) {
      const loaded: LoadedImage = preloader.getFile(graphFile)
      width = loaded.width
      height = loaded.width
    }
    filterChain.size = { width, height }


    if (streaming) graphFile.options!.re = ''

    return { graphFiles: [graphFile], link: this }
  }

  private graphFile(editing?: boolean): GraphFile {
    const { definition } = this
    const file = definition.preloadableSource(editing)
    assertPopulatedString(file, editing ? 'url' : 'source')

    const graphFile: GraphFile = {
      input: true, options: {}, type: LoadType.Image, file, definition
    }
    return graphFile
  }

  intrinsicDimensions(): Dimensions { return this.size }


  private setptsFilter: Filter

  _size?: Dimensions
  get size(): Dimensions { return this._size! }
  set size(value: Dimensions) { this._size = value }

  svgContent(filterChain: TrackPreview, dimensions: Dimensions): SvgContent {

    // TODO: consider removing method from definition
    return this.definition.svgContent(filterChain, dimensions)
  }
}
