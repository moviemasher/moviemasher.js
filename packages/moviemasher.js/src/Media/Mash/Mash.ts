import { UnknownRecord } from "../../Types/Core"
import { PreviewItems } from "../../Helpers/Svg/Svg"
import { AVType, MashType } from "../../Setup/Enums"
import { Time, Times } from "../../Helpers/Time/Time"
import { Clip, Clips } from "./Track/Clip/Clip"
import { TimeRange } from "../../Helpers/Time/Time"
import { Track, TrackObject } from "./Track/Track"
import { isArray, isObject } from "../../Utility/Is"

import { errorThrow } from "../../Helpers/Error/ErrorFunctions"
import { Request } from "../../Helpers/Request/Request"
import { Propertied } from "../../Base/Propertied"
import { Effect } from "../Effect/Effect"
import { Media, MediaObject, MediaObjects } from "../Media"
import { EncodingObjects, Encodings } from "../../Plugin/Encode/Encoding/Encoding"
import { Size } from "../../Utility/Size"
import { PreloadOptions } from "../../Base/Code"
import { Emitter } from "../../Helpers/Emitter"
import { MediaCollection } from "./MediaCollection/MediaCollection"
import { Masher, MashingType } from "../../Plugin/Masher/Masher"
import { Selectable } from "../../Plugin/Masher/Selectable"
import { AudioPreview } from "../../Plugin/Masher/Preview/AudioPreview/AudioPreview"

export type Movable = Effect 
export type Movables = Movable[]

export interface MashMediaContent {
  color?: string
  quantize?: number
  tracks?: TrackObject[]
  media?: MediaObjects
}

export interface MashMediaRequest extends Request {
  response?: MashMediaContent | undefined
}

export interface MashMediaObject extends MediaObject {
  request?: MashMediaRequest
  encodings?: EncodingObjects
}

export interface MashMasherArgs {
  mediaCollection: MediaCollection
  masher: Masher
  buffer: number
  gain: number
  loop: boolean
  emitter: Emitter
  size: Size
}
export interface MashMediaArgs extends MashMediaObject {
  
}
export interface MashAndMediaObject extends MashMediaObject {
  media: MediaObjects
}
export const isMashAndMediaObject = (value: any): value is MashAndMediaObject => {
  return isObject(value) && "media" in value && isArray(value.media)
}

/**
 * @category Media
 */
export interface MashMedia extends Media, Selectable {
  addClipToTrack(clip : Clip | Clips, trackIndex? : number, insertIndex? : number, frame? : number) : void
  addTrack(object?: TrackObject): Track
  buffer: number
  changeTiming(propertied: Propertied, property: string, value : number) : void
  clearPreview(): void
  clips: Clip[]
  clipsInTimeOfType(time: Time, avType?: AVType): Clip[]
  color: string
  composition: AudioPreview
  definitionIds: string[]
  destroy(): void
  draw() : void
  drawnTime? : Time
  duration: number
  editor: Masher
  emitter?: Emitter
  encodings: Encodings
  endTime: Time
  frame: number
  frames: number
  gain: number
  imageSize: Size
  kind: MashingType
  loading: boolean
  loadPromise(args?: PreloadOptions): Promise<void>
  loop: boolean
  media: MediaCollection
  paused: boolean
  previewItemsPromise(editor?: Masher): Promise<PreviewItems>
  putPromise(): Promise<void>
  quantize: number
  reload(): Promise<void> | undefined
  removeClipFromTrack(clip : Clip | Clips) : void
  removeTrack(index?: number): void
  request: MashMediaRequest

  seekToTime(time: Time): Promise<void> | undefined
  time: Time
  timeRange: TimeRange
  timeRanges(avType: AVType, startTime?: Time): Times
  toJSON(): UnknownRecord
  tracks: Track[]
  type: MashType
}
export const isMashMedia = (value: any): value is MashMedia => {
  return isObject(value) && "composition" in value
}
export function assertMashMedia(value: any, name?: string): asserts value is MashMedia {
  if (!isMashMedia(value)) errorThrow(value, "MashMedia", name)
}
