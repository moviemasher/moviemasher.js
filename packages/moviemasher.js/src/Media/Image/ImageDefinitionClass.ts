import { CanvasVisibleSource, SvgContent, LoadedImage } from "../../declarations"
import { Dimensions } from "../../Setup/Dimensions"
import { GraphFile, GraphFiles, GraphFileArgs } from "../../MoveMe"
import { DefinitionType, LoadType } from "../../Setup/Enums"
import { Image, ImageDefinition, ImageObject } from "./Image"
import { ImageClass } from "./ImageClass"
import { VisibleDefinitionMixin } from "../../Mixin/Visible/VisibleDefinitionMixin"
import { ClipDefinitionMixin } from "../../Mixin/Clip/ClipDefinitionMixin"
import { Time } from "../../Helpers/Time/Time"
import { Loader } from "../../Loader/Loader"
import { Errors } from "../../Setup/Errors"
import { PreloadableDefinitionMixin } from "../../Mixin/Preloadable/PreloadableDefinitionMixin"
import { DefinitionBase } from "../../Definition/DefinitionBase"
import { UpdatableDimensionsDefinitionMixin } from "../../Mixin/UpdatableDimensions/UpdatableDimensionsDefinitionMixin"
import { FilterDefinition } from "../../Filter/Filter"
import { filterDefinitionFromId } from "../../Filter/FilterFactory"
import { NamespaceSvg } from "../../Setup/Constants"
import { TrackPreview } from "../../Editor/Preview/TrackPreview/TrackPreview"
import { ContentDefinitionMixin } from "../../Content/ContentDefinitionMixin"
import { ContainerDefinitionMixin } from "../../Container/ContainerDefinitionMixin"
import { assertTrue } from "../../Utility/Is"

const ImageDefinitionWithContainer = ContainerDefinitionMixin(DefinitionBase)
const ImageDefinitionWithContent = ContentDefinitionMixin(ImageDefinitionWithContainer)
const ImageDefinitionWithPreloadable = PreloadableDefinitionMixin(ImageDefinitionWithContent)
const ImageDefinitionWithUpdatable = UpdatableDimensionsDefinitionMixin(ImageDefinitionWithPreloadable)
export class ImageDefinitionClass extends ImageDefinitionWithUpdatable implements ImageDefinition {
  constructor(...args: any[]) {
    super(...args)
    this.setptsFilterDefinition = filterDefinitionFromId('setpts')
  }

  graphFiles(args: GraphFileArgs): GraphFiles {
    const graphFiles: GraphFiles = []
    const graphFile = this.preloadableFile(args)
    if (graphFile) graphFiles.push(graphFile)
    return graphFiles
  }

  instanceFromObject(object: ImageObject = {}) : Image {
    return new ImageClass(this.instanceArgs(object))
  }

  loadedVisible(preloader: Loader, quantize: number, time: Time): CanvasVisibleSource | undefined {
    const filesArgs: GraphFileArgs = { visible: true, editing: true, time: time, quantize }
    const file = this.preloadableFile(filesArgs)
    if (!file || !preloader.loadedFile(file)) return

    return preloader.getFile(file)
  }

  loadType = LoadType.Image

  private preloadableFile(args: GraphFileArgs): GraphFile | undefined {
    const { editing, visible } = args
    if (!visible) return

    const file = this.preloadableSource(editing)
    if (!file) throw Errors.invalid.url + this.id + ' preloadableFile'

    const graphFile: GraphFile = {
      type: this.loadType, file, definition: this, input: true, localId: LoadType.Image,
    }
    return graphFile
  }

  svgContent(filterChain: TrackPreview, dimensions?: Dimensions): SvgContent {
    const { filterGraph } = filterChain
    const { preloader, size } = filterGraph
    const file = this.preloadableFile(filterGraph)!

    const loaded: LoadedImage = preloader.getFile(file)
    assertTrue(loaded, 'LoadedImage')

    const { src: href, width: loadedWidth, height: loadedHeight } = loaded
    const imageElement = globalThis.document.createElementNS(NamespaceSvg, 'image')
    imageElement.setAttribute('id', `image-${this.id}`)
    imageElement.setAttribute('href', href)

    const { width, height } = dimensions || size
    const scaleWidth = width / loadedWidth
    const scaleHeight = height / loadedHeight
    if (scaleWidth > scaleHeight) imageElement.setAttribute('width', String(width))
    else imageElement.setAttribute('height', String(height))

    return imageElement
  }

  setptsFilterDefinition: FilterDefinition

  type = DefinitionType.Image
}
