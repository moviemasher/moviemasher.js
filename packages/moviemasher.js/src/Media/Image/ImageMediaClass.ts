import { ClientImage } from "../../ClientMedia/ClientMedia"
import { ImageType } from "../../Setup/Enums"
import { Image, ImageMedia, ImageObject } from "./Image"
import { ImageClass } from "./ImageClass"
import { UpdatableSizeDefinitionMixin } from "../../Mixin/UpdatableSize/UpdatableSizeDefinitionMixin"
import { ContentDefinitionMixin } from "../Content/ContentDefinitionMixin"
import { ContainerDefinitionMixin } from "../Container/ContainerDefinitionMixin"
import { TweenableDefinitionMixin } from "../../Mixin/Tweenable/TweenableDefinitionMixin"
import { assertSizeAboveZero, Size, sizeCover } from "../../Utility/Size"
import { MediaBase } from "../MediaBase"
import { PreloadArgs } from "../../Base/Code"
import { requestImagePromise } from "../../Helpers/Request/RequestFunctions"
import { centerPoint, RectOptions } from "../../Utility/Rect"
import { svgImagePromiseWithOptions, svgSvgElement } from "../../Helpers/Svg/SvgFunctions"
import { errorThrow } from "../../Helpers/Error/ErrorFunctions"
import { isDefiniteError } from "../../Utility/Is"

const ImageMediaWithTweenable = TweenableDefinitionMixin(MediaBase)
const ImageMediaWithContainer = ContainerDefinitionMixin(ImageMediaWithTweenable)
const ImageMediaWithContent = ContentDefinitionMixin(ImageMediaWithContainer)
const ImageMediaWithUpdatable = UpdatableSizeDefinitionMixin(ImageMediaWithContent)
export class ImageMediaClass extends ImageMediaWithUpdatable implements ImageMedia {
  definitionIcon(size: Size): Promise<SVGSVGElement> | undefined {
    const transcoding = this.preferredTranscoding(ImageType) 

    return requestImagePromise(transcoding.request).then(orError => {
      if (isDefiniteError(orError)) return errorThrow(orError.error)

      const { data: clientImage } = orError
      assertSizeAboveZero(clientImage)

      const { width, height, src } = clientImage
      const inSize = { width, height }
      const coverSize = sizeCover(inSize, size, true)
      const outRect = { ...coverSize, ...centerPoint(size, coverSize) }
      const options: RectOptions = { ...outRect }
      return svgImagePromiseWithOptions(src, options).then(svgImage => {
        return svgSvgElement(size, svgImage)
      })
    })
  }

  instanceFromObject(object: ImageObject = {}) : Image {
    return new ImageClass(this.instanceArgs(object))
  }


  loadPromise(args: PreloadArgs): Promise<void> {
    const { editing, visible } = args
    if (!visible) return Promise.resolve()

    const { loadedImage } = this
    if (loadedImage) return Promise.resolve()

    const transcoding = editing ? this.preferredTranscoding(ImageType) : this
    
    const { request } = transcoding
    return requestImagePromise(request).then(orError => {
      if (isDefiniteError(orError)) return errorThrow(orError.error)

      const { data: clientImage } = orError
      this.loadedImage = clientImage
    })
  }
  
  loadedImage?: ClientImage 
  
  type = ImageType 
}
