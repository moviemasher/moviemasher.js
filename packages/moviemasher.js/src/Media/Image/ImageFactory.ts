import { ImageMediaClass } from "./ImageMediaClass"
import { Image, ImageMedia, ImageMediaObject, ImageObject } from "./Image"
import { assertPopulatedString } from "../../Utility/Is"
import { MediaFactories } from "../MediaFactories"
import { ImageType } from "../../Setup/Enums"

import defaultContent from "../../MediaObjects/content/default.json"

import defaultContainer from "../../MediaObjects/container/default.json"
import heartContainer from "../../MediaObjects/container/heart.json"
import cloudContainer from "../../MediaObjects/container/cloud.json"
import appleContainer from "../../MediaObjects/container/apple.json"
import starburstContainer from "../../MediaObjects/container/starburst.json"
import roundedRectContainer from "../../MediaObjects/container/rounded-rect.json"
import fireContainer from "../../MediaObjects/container/fire.json"
import flagContainer from "../../MediaObjects/container/flag.json"
import ovalContainer from "../../MediaObjects/container/oval.json"
import mmContainer from "../../MediaObjects/container/mm.json"
import { ShapeContainerDefinitionClass } from "../Container/ShapeContainer/ShapeContainerDefinitionClass"
import { DefaultContainerId } from "../Container/Container"
import { MediaDefaults } from "../MediaDefaults"
import { ColorContentDefinitionClass } from "../Content/ColorContent/ColorContentDefinitionClass"
import { DefaultContentId } from "../Content/Content"

export const imageDefinition = (object : ImageMediaObject) : ImageMedia => {
  const { id } = object
  assertPopulatedString(id, 'imageDefinition id')

  return new ImageMediaClass(object)
}

export const imageDefinitionFromId = (id : string) : ImageMedia => {
  const definition = MediaDefaults[ImageType].find(definition => 
    definition.id === id
  )
  if (definition) return definition as ImageMedia

  
  return imageDefinition({ id })
}

export const imageInstance = (object : ImageObject) : Image => {
  const { mediaId: id, definition: defined } = object
  const definition = defined || imageDefinitionFromId(id!)
  const instance = definition.instanceFromObject(object)
  return instance
}

export const imageFromId = (id : string) : Image => {
  return imageInstance({ id })
}


MediaFactories[ImageType] = imageDefinition

MediaDefaults[ImageType].push(
  new ColorContentDefinitionClass({id: DefaultContentId, ...defaultContent}),
  new ShapeContainerDefinitionClass({ id: DefaultContainerId, ...defaultContainer }),
  new ShapeContainerDefinitionClass(roundedRectContainer),
  new ShapeContainerDefinitionClass(ovalContainer),
  new ShapeContainerDefinitionClass(starburstContainer),
  new ShapeContainerDefinitionClass(heartContainer),
  new ShapeContainerDefinitionClass(cloudContainer),
  new ShapeContainerDefinitionClass(fireContainer),
  new ShapeContainerDefinitionClass(flagContainer),
  new ShapeContainerDefinitionClass(mmContainer),
  new ShapeContainerDefinitionClass(appleContainer),

)

