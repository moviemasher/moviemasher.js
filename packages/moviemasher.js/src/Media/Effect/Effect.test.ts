import { DefinitionType, EditType, TrackType } from "../../Setup/Enums"
import { expectCanvasAtTime } from "../../../../../dev/test/Utilities/expectMashSvg"
import { editorInstance } from "../../Editor/EditorFactory"
import { JestPreloader } from "../../../../../dev/test/Utilities/JestPreloader"
import { assertMash } from "../../Edited/Mash/Mash"
import { effectDefinitionFromId } from "./EffectFactory"
import { imageDefinition } from "../Image/ImageFactory"
import { ImageDefinitionObject } from "../Image/Image"
import { Defined } from "../../Base/Defined"

describe("Effect", () => {
  describe("ChromaKey", () => {
    test("returns expected context", async () => {
      const matteDefinitionObject: ImageDefinitionObject = {
        id: 'text-matte', url: '../shared/image/green-text-on-white.png',
        type: DefinitionType.Image
      }
      const imageDefinitionObject: ImageDefinitionObject = {
        id: 'cable-image', url: '../shared/image/cable.jpg',
        type: DefinitionType.Image
      }
      const editor = editorInstance({ editType: EditType.Mash, preloader: new JestPreloader() })
      editor.load({mash: {}})
      const { edited } = editor
      assertMash(edited)
      Defined.define(matteDefinitionObject, imageDefinitionObject)
      editor.imageSize = { width: 640, height: 480 }
      editor.addTrack(TrackType.Video)
      expect(edited.tracks.length).toBe(3)
      const matteDefinition = imageDefinition(matteDefinitionObject)
      const cableDefinition = imageDefinition(imageDefinitionObject)

      expect(cableDefinition.preloadableSource(true)).toBe(imageDefinitionObject.url)
      expect(matteDefinition.preloadableSource(true)).toBe(matteDefinitionObject.url)
      const effectDefinition = effectDefinitionFromId("com.moviemasher.effect.chromakey")
      const matteImage = matteDefinition.instanceFromObject()
      const imageImage = cableDefinition.instanceFromObject()
      await editor.addClip(imageImage)
      await editor.addClip(matteImage, 0, 1)
      editor.select(matteImage)
      await editor.addEffect(effectDefinition.instanceFromObject())
      expect(edited.trackOfTypeAtIndex(TrackType.Video, 0).clips).toEqual([imageImage])
      expect(edited.trackOfTypeAtIndex(TrackType.Video, 1).clips).toEqual([matteImage])
      expectCanvasAtTime(editor)
    })
  })
})
