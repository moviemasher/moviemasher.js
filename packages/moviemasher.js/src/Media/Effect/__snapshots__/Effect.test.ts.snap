// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Effect ChromaKey returns expected context: Unlabeled Mash 1`] = `"<svg width=\\"640\\" height=\\"480\\" viewBox=\\"0 0 640 480\\"><svg width=\\"320\\" height=\\"240\\" viewBox=\\"0 0 320 240\\"><image id=\\"image-1\\" width=\\"320\\" height=\\"240\\" href=\\"http://localhost/shared/image/cable.jpg\\" style=\\"filter:url(#filter-3);\\"></image><filter id=\\"filter-3\\"><feOffset id=\\"com.moviemasher.filter.overlay\\" dx=\\"0\\" dy=\\"0\\"></feOffset></filter></svg><svg width=\\"774\\" height=\\"540\\" viewBox=\\"0 0 774 540\\"><image id=\\"image-2\\" width=\\"774\\" height=\\"540\\" href=\\"http://localhost/shared/image/green-text-on-white.png\\" style=\\"filter:url(#filter-4);\\"></image><filter id=\\"filter-4\\"><feOffset id=\\"com.moviemasher.filter.overlay\\" dx=\\"0\\" dy=\\"0\\"></feOffset></filter></svg></svg>"`;
